xarray/tutorial.py

```diff
@@ -6,23 +6,129 @@
 
 """
 import hashlib
-import os as _os
-from urllib.request import urlretrieve
+import pathlib
+import shutil
+import tempfile
+from contextlib import contextmanager, suppress
 
 import numpy as np
+import requests
 
 from .backends.api import open_dataset as _open_dataset
+from .backends.rasterio_ import open_rasterio as _open_rasterio
 from .core.dataarray import DataArray
 from .core.dataset import Dataset
+from .vendor import appdirs
 
-_default_cache_dir = _os.sep.join(("~", ".xarray_tutorial_data"))
+_default_cache_dir = pathlib.Path(appdirs.user_cache_dir("xarray_tutorial_data"))
 
 
-def file_md5_checksum(fname):
-    hash_md5 = hashlib.md5()
-    with open(fname, "rb") as f:
-        hash_md5.update(f.read())
-    return hash_md5.hexdigest()
+def check_md5sum(content, checksum):
+    md5 = hashlib.md5()
+    md5.update(content)
+    md5sum = md5.hexdigest()
+
+    return md5sum == checksum
+
+
+# based on https://stackoverflow.com/a/29491523
+@contextmanager
+def open_atomic(path, mode=None):
+    folder = path.parent
+    prefix = f".{path.name}"
+
+    try:
+        f = tempfile.NamedTemporaryFile(dir=folder, prefix=prefix, delete=False)
+        temporary_path = pathlib.Path(f.name)
+
+        yield f
+
+        temporary_path.rename(path)
+    except Exception:
+        # switch to path.unlink(missing_ok=True) for Python 3.8+
+        with suppress(FileNotFoundError):
+            temporary_path.unlink()
+        raise
+
+
+def download_to(url, path):
+    # based on https://stackoverflow.com/a/39217788
+    with open_atomic(path, mode="wb") as f:
+        with requests.get(url, stream=True) as r:
+            if r.status_code != 200:
+                raise OSError(f"download failed: {r.reason}")
+
```

Comment:
would this suffice instead?
```suggestion
           r.raise_for_status()
```

Reply:
I don't think so? The [docs](https://requests.readthedocs.io/en/master/user/quickstart/#response-status-codes) say that would only raise for `4xx` or `5xx`, so this is not strictly the same. We could replace the `200` with `requests.codes.ok`.